controllers/pipelinerun_controller.go

```diff
@@ -0,0 +1,203 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/shipwright-io/build/pkg/apis/build/v1alpha1"
+	"github.com/shipwright-io/triggers/pkg/filter"
+	"github.com/shipwright-io/triggers/pkg/inventory"
+	"github.com/shipwright-io/triggers/test/stubs"
+
+	"github.com/go-logr/logr"
+	tknv1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// PipelineRunReconciler reconciles PipelineRun objects that may have triggers configured to generate
+// a BuildRun based on the Pipeline state.
+type PipelineRunReconciler struct {
+	client.Client                 // kubernetes client
+	Scheme        *runtime.Scheme // shared scheme
+	Clock                         // local clock instance
+
+	buildInventory *inventory.Inventory // local build triggers database
+}
+
+//+kubebuilder:rbac:groups=shipwright.io,resources=builds,verbs=get;list;watch
+//+kubebuilder:rbac:groups=shipwright.io,resources=buildruns,verbs=create;get;list;update;watch
+//+kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns,verbs=get;list;update;watch
+
+// createBuildRun handles the actual BuildRun creation, uses the informed PipelineRun instance to
+// establish ownership. Only returns the created object name, and error.
+func (r *PipelineRunReconciler) createBuildRun(
+	ctx context.Context,
+	pipelineRun *tknv1beta1.PipelineRun,
+	buildName string,
+) (string, error) {
+	br := v1alpha1.BuildRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    pipelineRun.GetNamespace(),
+			GenerateName: fmt.Sprintf("%s-", pipelineRun.GetName()),
+			Labels: map[string]string{
+				filter.OwnedByPipelineRunLabelKey: pipelineRun.GetName(),
+			},
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: stubs.TektonAPIv1beta1,
+				Kind:       "PipelineRun",
+				Name:       pipelineRun.GetName(),
+				UID:        pipelineRun.GetUID(),
+			}},
```

Comment:
Hm. The scenarios imo can be very different and whether there should be that strong relationship, I am not sure. I would though expect to see above for the custom task integration where the BuildRun really becomes part of the PipelineRun. The trigger on pipelinerun status is imo more loosly coupling these two artifacts. We should still have a label, but that should maybe be "triggered-by" rather than an ownership. Wdyt?

Reply:
As a initial approach, I think the strong relationship makes sense because it keeps the cluster clean, when the `PipelineRun` gets removed the objects triggered will also get removed.

The rationale is based on the fact that both Shipwright and Tekton produce ephemeral objects, these are only useful when the build process is running and afterwards should not leave instances behind.